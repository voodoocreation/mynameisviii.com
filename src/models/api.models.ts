import { arrayToAssoc } from "../transformers/transformData";

interface IWithSlug {
  slug: string;
}

export type TLastEvaluatedKey<K extends string> = Record<
  K | "isActive" | "slug",
  string
>;

export interface IRawDynamoResponse<T extends {}, K extends string> {
  items: T[];
  lastEvaluatedKey?: TLastEvaluatedKey<K>;
}

export interface IDynamoResponse<T extends IWithSlug, K extends string> {
  readonly items: Record<string, T>;
  readonly lastEvaluatedKey?: TLastEvaluatedKey<K>;
}

export interface IRawS3Response<T extends {}> {
  isTruncated?: boolean;
  items: T[];
}

export interface IS3Response<T extends IWithSlug> {
  readonly isTruncated: boolean;
  readonly items: Record<string, T>;
}

export const dynamoResponse = <T extends IWithSlug, K extends string>(
  options: IRawDynamoResponse<T, K>
): IDynamoResponse<T, K> => ({
  items: arrayToAssoc(options.items, "slug"),
  lastEvaluatedKey: options.lastEvaluatedKey
});

export const s3Response = <T extends IWithSlug>(
  options: IRawS3Response<T>
): IS3Response<T> => ({
  isTruncated: !!options.isTruncated,
  items: arrayToAssoc(options.items, "slug")
});
