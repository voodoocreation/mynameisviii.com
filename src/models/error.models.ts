export interface IError {
  readonly message: string;
  readonly status: number;
}

export const error = (options: Partial<IError> = {}): IError => ({
  message: options.message || "An error has occurred",
  status: options.status || 500,
});
