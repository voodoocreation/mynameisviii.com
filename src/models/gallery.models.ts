import dayjs from "dayjs";

export interface IRawGallery {
  description?: string;
  imageUrl?: string;
  images?: Array<Partial<IGalleryImage>>;
  modifiedAt?: string;
  slug?: string;
  title?: string;
}

export interface IGallery {
  readonly description?: string;
  readonly imageUrl: string;
  readonly images?: IGalleryImage[];
  readonly modifiedAt: string;
  readonly slug: string;
  readonly title: string;
}

export interface IGalleryImage {
  readonly imageUrl: string;
  readonly modifiedAt: string;
}

export const galleryImage = (
  options: Partial<IGalleryImage> = {}
): IGalleryImage => ({
  imageUrl: options.imageUrl || "",
  modifiedAt: options.modifiedAt || dayjs().toISOString()
});

export const gallery = (options: IRawGallery = {}): IGallery => ({
  description: options.description,
  imageUrl: options.imageUrl || "",
  images: options.images ? options.images.map(galleryImage) : undefined,
  modifiedAt: options.modifiedAt || dayjs().toISOString(),
  slug: options.slug || "",
  title: options.title || ""
});
