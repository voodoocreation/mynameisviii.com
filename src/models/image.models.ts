export interface IImage {
  readonly imageUrl: string;
  readonly title: string;
}

export const image = (options: Partial<IImage> = {}): IImage => ({
  imageUrl: options.imageUrl || "",
  title: options.title || ""
});
