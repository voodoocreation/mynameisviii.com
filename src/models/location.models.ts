import { TYPE } from "../constants/location.constants";

export interface ILatLng {
  readonly lat: number;
  readonly lng: number;
}

export interface ILocation {
  readonly address?: string;
  readonly latLng?: ILatLng;
  readonly name: string;
  readonly type: TYPE;
  readonly url?: string;
}

export const location = (options: Partial<ILocation> = {}): ILocation => ({
  address: options.address,
  latLng: options.latLng,
  name: options.name || "",
  type: options.type || TYPE.MUSIC_VENUE,
  url: options.url,
});
