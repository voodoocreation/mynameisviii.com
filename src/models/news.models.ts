import dayjs from "dayjs";

import { BOOLEAN } from "../constants/api.constants";
import { TYPE } from "../constants/news.constants";

export interface INewsArticleAction {
  route?: string;
  url?: string;
  text: string;
}

export interface IRawNewsArticle {
  action?: INewsArticleAction;
  author?: string;
  content?: string;
  createdAt?: string;
  excerpt?: string;
  imageUrl?: string;
  isActive?: BOOLEAN;
  ogImageUrl?: string;
  slug?: string;
  title?: string;
  type?: TYPE;
}

export interface INewsArticle {
  readonly action?: INewsArticleAction;
  readonly author: string;
  readonly content: string;
  readonly createdAt: string;
  readonly excerpt: string;
  readonly imageUrl: string;
  readonly isActive: boolean;
  readonly ogImageUrl: string;
  readonly slug: string;
  readonly title: string;
  readonly type: TYPE;
}

export const newsArticle = (options: IRawNewsArticle = {}): INewsArticle => ({
  action: options.action,
  author: options.author || "",
  content: options.content || "",
  createdAt: options.createdAt || dayjs().toISOString(),
  excerpt: options.excerpt || "",
  imageUrl: options.imageUrl || "",
  isActive: (options.isActive && options.isActive === BOOLEAN.TRUE) || false,
  ogImageUrl: options.ogImageUrl || "",
  slug: options.slug || "",
  title: options.title || "",
  type: options.type || TYPE.NEWS
});
