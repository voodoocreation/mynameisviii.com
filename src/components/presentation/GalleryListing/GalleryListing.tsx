import cn from "classnames";
import * as React from "react";

import { IGallery } from "../../../models/root.models";
import Schema from "../../schema/Gallery";
import Image from "../Image/Image";
import Link from "../Link/Link";

import "./GalleryListing.scss";

interface IState {
  isRendered: boolean;
}

interface IProps extends IGallery {
  onLoad?: () => void;
}

export default class GalleryListing extends React.Component<IProps, IState> {
  public readonly state = {
    isRendered: false
  };

  public componentDidMount() {
    if (!this.props.imageUrl) {
      this.onLoad();
    }
  }

  public render() {
    const { onLoad, ...gallery } = this.props;
    const { isRendered } = this.state;

    return (
      <article className={cn("GalleryListing", { isRendered })}>
        <Link route={`/galleries/${gallery.slug}`}>
          <header className="GalleryListing--header">
            <h2>{gallery.title}</h2>
          </header>

          <Image
            alt={gallery.title}
            className="GalleryListing--image"
            src={gallery.imageUrl}
            onLoad={this.onLoad}
          />
        </Link>

        <Schema {...gallery} />
      </article>
    );
  }

  private onLoad = () => {
    this.setState({
      isRendered: true
    });

    if (this.props.onLoad) {
      this.props.onLoad();
    }
  };
}
