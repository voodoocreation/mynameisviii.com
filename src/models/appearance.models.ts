import dayjs from "dayjs";

import { BOOLEAN } from "../constants/api.constants";
import { STATUS, TYPE } from "../constants/appearance.constants";
import { IImage, image } from "./image.models";
import { ILocation, location } from "./location.models";
import { IOffer, offer } from "./offer.models";
import { IOrganization, organization } from "./organization.models";
import { IPerformer, IRawPerformer, performer } from "./performer.models";

export interface IRawAppearance {
  acts?: Array<Partial<IRawPerformer>>;
  audience?: string;
  description?: string;
  finishingAt?: string;
  imageUrl?: string;
  images?: IImage[];
  isActive?: BOOLEAN;
  location?: Partial<ILocation>;
  ogImageUrl?: string;
  organizer?: Partial<IOrganization>;
  rsvpUrl?: string;
  sales?: Array<Partial<IOffer>>;
  slug?: string;
  startingAt?: string;
  status?: STATUS;
  title?: string;
  type?: TYPE;
}

export interface IAppearance {
  readonly acts: IPerformer[];
  readonly audience?: string;
  readonly description: string;
  readonly finishingAt: string;
  readonly imageUrl: string;
  readonly images: IImage[];
  readonly isActive: boolean;
  readonly location: ILocation;
  readonly ogImageUrl: string;
  readonly organizer: IOrganization;
  readonly rsvpUrl?: string;
  readonly sales: IOffer[];
  readonly slug: string;
  readonly startingAt: string;
  readonly status: STATUS;
  readonly title: string;
  readonly type: TYPE;
}

export const appearance = (options: IRawAppearance = {}): IAppearance => ({
  acts: options.acts ? options.acts.map(performer) : [],
  audience: options.audience,
  description: options.description || "",
  finishingAt: options.finishingAt || dayjs().toISOString(),
  imageUrl: options.imageUrl || "",
  images: options.images ? options.images.map(image) : [],
  isActive: (options.isActive && options.isActive === BOOLEAN.TRUE) || false,
  location: location(options.location),
  ogImageUrl: options.ogImageUrl || "",
  organizer: organization(options.organizer),
  rsvpUrl: options.rsvpUrl,
  sales: options.sales ? options.sales.map(offer) : [],
  slug: options.slug || "",
  startingAt: options.startingAt || dayjs().toISOString(),
  status: options.status || STATUS.SCHEDULED,
  title: options.title || "",
  type: options.type || TYPE.MUSIC,
});
