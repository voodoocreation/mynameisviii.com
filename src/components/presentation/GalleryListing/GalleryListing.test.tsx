import { gallery } from "../../../models/root.models";
import ComponentTester from "../../../utilities/ComponentTester";
import GalleryListing from "./GalleryListing";

const component = new ComponentTester(GalleryListing).withDefaultProps(
  gallery({
    description: "Description",
    modifiedAt: "2017-10-10T18:00:00",
    slug: "test-1",
    title: "Title"
  })
);

describe("[presentation] <GalleryListing />", () => {
  describe("when imageUrl is defined", () => {
    const { props, wrapper } = component
      .withProps({
        imageUrl: "Image URL",
        onLoad: jest.fn()
      })
      .mount();

    it("doesn't render with isRendered class instantly", () => {
      expect(wrapper.find(".isRendered")).toHaveLength(0);
    });

    it("doesn't call onLoad prop instantly", () => {
      expect(props.onLoad).toHaveBeenCalledTimes(0);
    });

    it("loads the image", () => {
      wrapper.find("Image").simulate("load");
    });

    it("adds isRendered class", () => {
      expect(wrapper.find(".isRendered")).toHaveLength(1);
    });

    it("calls onLoad prop", () => {
      expect(props.onLoad).toHaveBeenCalledTimes(1);
    });

    it("matches snapshot", () => {
      expect(wrapper.render()).toMatchSnapshot();
    });
  });

  describe("when imageUrl isn't defined", () => {
    const { wrapper } = component.withProps({ imageUrl: undefined }).mount();

    it("renders with isRendered class instantly", () => {
      expect(wrapper.find(".isRendered")).toHaveLength(1);
    });
  });
});
