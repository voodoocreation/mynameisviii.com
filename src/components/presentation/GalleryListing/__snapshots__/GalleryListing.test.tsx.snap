// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[presentation] <GalleryListing /> when imageUrl is defined matches snapshot 1`] = `
<article
  class="GalleryListing isRendered"
>
  <a
    href="/galleries/test-1"
  >
    <header
      class="GalleryListing--header"
    >
      <h2>
        Title
      </h2>
    </header>
    <figure
      class="Image GalleryListing--image isLoading"
    >
      <div
        class="Image--wrapper"
      >
        <img
          alt="Title"
          src="Image URL"
        />
        <span
          class="Loader"
        >
          <div
            class="Loader--animation"
          >
            <span />
            <span />
            <span />
          </div>
        </span>
      </div>
    </figure>
  </a>
  <script
    type="application/ld+json"
  >
    {"@id":"http://localhost:80/galleries/test-1","@type":"ImageGallery","dateModified":"2017-10-10T18:00:00","description":"Description","name":"Title","primaryImageOfPage":"Image URL","thumbnailUrl":"Image URL","@context":"http://schema.org/"}
  </script>
</article>
`;
